### PR TITLE
Make latency prediction server multi threaded compatable

### PR DESCRIPTION
**Pull request by kaushikmitr**
This pull request introduces several improvements across configuration, deployment, and testing for the latency predictor service. The most notable changes include making the number of Uvicorn worker processes configurable in the Docker deployment, updating stress test batch sizes for bulk prediction, and adding project ownership metadata.

**Deployment and configuration improvements:**

* Made the number of Uvicorn worker processes configurable via the `UVICORN_WORKERS` environment variable in the `latencypredictor/Dockerfile-prediction`, and updated the container command to use this variable.

* Make `latencypredictor/prediction_server.py` multi thread compatible. 

**Testing enhancements:**

* Increased the batch sizes in the bulk prediction stress test in `latencypredictor/test_dual_server_client.py` to `[25, 50, 100]` for more robust testing.

**Project metadata:**

* Added an `OWNERS` file to `latencypredictor` specifying project approvers.

### CODE REVIEW

**Comment by kaushikmitr** [comment]
@ahg-g @Gregory-Pereira @BenjaminBraunDev results are good by allowing multi threads. We still need to implement throttling though.
### **Performance Test Results**

**Setup Configuration**

| Parameter | Value |
| :--- | :--- |
| **Machine** |  GCE e2-highcpu-32 — 32 vCPUs, 32 GB RAM (28 CPUs allocated to prediction server)  |
| **Prediction server replicas** | 1 |
| **`UVICORN_WORKERS`** | 28 |
| **`OMP_NUM_THREADS`** | 1 |
| **Endpoints per request** | 100 |
| **Test duration** | 300s |
| **Training batch size (executed in parallel)** | 1,000 entries every 500ms |
| **Model type** | XGBoost |

<br>

**Load Test Results**

| Metric | 1000 QPS | 1500 QPS | 2000 QPS |
| :--- | :--- | :--- | :--- |
| **Success rate** | 100% | 100% | 96.4% |
| **Total predictions** | 29.7M | 36.8M | N/A *(aborted)* |
| **Avg latency** | 17ms | 33ms | 300-480ms |
| **P50** | 16ms | 30ms | — |
| **P99** | 36ms | 81ms | — |
| **P999** | 81ms | 113ms | — |

**Comment by ahg-g** [comment]
> Prediction server replicas | 1

Do you mean a single sidecar?

> UVICORN_WORKERS | 28

How did we determine this number? what happens if we have more? 

> OMP_NUM_THREADS | 1

Why 1? how does this relate to UVICORN_WORKERS?

**Comment by ahg-g** [comment]
Pls lets avoid doing multiple major changes in a single PR, I can't tell which changes in the file related to which enhancement from the list in the description. 

> Made the number of Uvicorn worker processes configurable via the UVICORN_WORKERS environment variable in the latencypredictor/Dockerfile-prediction, and updated the container command to use this variable.

What was this set to before?

> In cmd/epp/runner/runner.go, added a call to PopulateControllerConfig to ensure the controller configuration is properly populated before starting the service, with error handling and logging for failures.

I don't see changes to any go files in this PR.

> Make latencypredictor/prediction_server.py multi thread compatible.

I see updates to the model loading logic, is that related? what form of parallelism did we have before? just the low level model prediction logic for a single bulk request and requests were processed sequentially?

**Comment by kaushikmitr** [comment]
> Pls lets avoid doing multiple major changes in a single PR, I can't tell which changes in the file related to which enhancement from the list in the description.
> 
> > Made the number of Uvicorn worker processes configurable via the UVICORN_WORKERS environment variable in the latencypredictor/Dockerfile-prediction, and updated the container command to use this variable.
> 
> What was this set to before?
> 
> > In cmd/epp/runner/runner.go, added a call to PopulateControllerConfig to ensure the controller configuration is properly populated before starting the service, with error handling and logging for failures.
> 
> I don't see changes to any go files in this PR.
> 
> > Make latencypredictor/prediction_server.py multi thread compatible.
> 
> I see updates to the model loading logic, is that related? what form of parallelism did we have before? just the low level model prediction logic for a single bulk request and requests were processed sequentially?



> Pls lets avoid doing multiple major changes in a single PR, I can't tell which changes in the file related to which enhancement from the list in the description.
> 
> > Made the number of Uvicorn worker processes configurable via the UVICORN_WORKERS environment variable in the latencypredictor/Dockerfile-prediction, and updated the container command to use this variable.
> 
> What was this set to before?
> 
> > In cmd/epp/runner/runner.go, added a call to PopulateControllerConfig to ensure the controller configuration is properly populated before starting the service, with error handling and logging for failures.
> 
> I don't see changes to any go files in this PR.
> 
> > Make latencypredictor/prediction_server.py multi thread compatible.
> 
> I see updates to the model loading logic, is that related? what form of parallelism did we have before? just the low level model prediction logic for a single bulk request and requests were processed sequentially?

sorry the commit description was misleading. There was an error in the description and I edited it now. This PR has only 1 change spread across 2 files, and updated a test to increase batch size during stress test. 
1. Made the number of Uvicorn worker processes configurable via the UVICORN_WORKERS environment variable in the latencypredictor/Dockerfile-prediction, and updated the container command to use this variable.

2. Make latencypredictor/prediction_server.py multi thread compatible. 
3. Updated the test to increase batch size to upto 100. This is not directly related just increases load. 

** i tried to squash my commits and i guess messed up something, let me fix it

**Comment by kaushikmitr** [comment]
Previously i had 1 Uvicorn worker (default) with XGBoost using all available CPUs per prediction (OpenMP default). So single  prediction at a time (due to the lock), using all cores for that one prediction via XGBoost's internal parallelism. 

Now i have 28 predictions concurrently with each prediction using only 1 thread. The latter seems to give better throughput

**Comment by kaushikmitr** [comment]
closed in favor of https://github.com/kubernetes-sigs/gateway-api-inference-extension/pull/2349
